### PR TITLE
[FEAT/44] Added Facebook OAuth2 To Authenticate Users

### PR DESCRIPTION
**Pull request by mmubarak0**
## Confirmed: The code has been verified as safe.🛩 ✅
Implemented an endpoint for user login via social media accounts. It manages all server-to-server communications required to authenticate the user.

> This code implements social authentication safely, ensuring no harm to the database model. It creates a separate entity named `OAuthUser`, which is associated with the user model through a relationship, leaving the `User` model unmodified.

## Related Issue:

- https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/55
​

## Description
The endpoint `/api/v1/auth/facebook-login` is created for user authentication via their Facebook account. It ensures a secure login by interfacing with the Facebook API, providing a seamless user experience.

## Testing

The endpoint has undergone testing via Postman and PyTest, The resulting images are provided below at the end.

#### To run test cases use the following command

```py
python -m pytest tests/v1/test_facebook_login.py
```

## Steps to test the endpoint

- first go to https://developers.facebook.com/tools/explorer
- Login with your developer account if asked.
- generate a new access token an pass
- use your newly created access token to test the endpoint 
- here is a curl command if prefer to use command line
```bash
curl -X POST -H "Content-Type: application/json" -d '{"access_token": "YOUR_ACCESS_TOKEN"}' http://example.com/api/v1/auth/facebook-login
```

## Types of changes

- [x] Integrate Facebook OAuth API.
- [x] Create backend endpoints to handle OAuth authentication ``` api/v1/auth/facebook-login ```
- [x] Secure storage of OAuth tokens.
- [x] Ensure payload/response objects as follow ``` { "message": string, "status_code": number } ```
- [x] Ensure compliance with Facebook’s OAuth policies.
- [x] All new and existing test are passed.

#### successful login

> 201 response: the user is login in for the first time
image
> 200 response: the is already registered on the database
image

image

#### missing token

> 422 response: the token is missing from the response but the body is passed correctly
image
> 422 response: the body has not been passed correctly
image

### CODE REVIEW

**Comment by Dev-wonderful** [comment]
Resolve confflicts

File: api/v1/routes/auth_facebook.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from api.v1.services.facebook import validate_facebook_token, create_db_user, create_oauth_user
+
+from api.v1.schemas.facebook import userToken
+
+# don't remove the following imports
+from api.v1.models.user import User
+from api.v1.models.oauth import OAuthUser
+from api.v1.models.profile import Profile
+from api.v1.models.org import Organization
+# -------------------------------
```
Comment (Dev-wonderful) [inline api/v1/routes/auth_facebook.py:26]: everything is in the `models/__init__.py` file, import from there `from models import User, Profile ....`

**Review by Dev-wonderful — CHANGES_REQUESTED** [review]

File: api/v1/routes/auth_facebook.py (hunk 1 above)
Comment (mmubarak0) [reply api/v1/routes/auth_facebook.py:26]: I'm working on it

File: api/v1/routes/auth_facebook.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            response = JsonResponseDict(
+                message="Invalid access token",
+                status_code=status.HTTP_401_UNAUTHORIZED,
+                error=COULD_NOT_VALIDATE_CRED,
+            )
+            return response
+        fb_user_id = fb_user[0]
+        fb_user_data = fb_user[1]
+
+        # retrieve the user from the database
```
[the remark this answers is not in the document]
Comment (mmubarak0) [reply api/v1/routes/auth_facebook.py]: have I did the right thing here?

**Comment by mmubarak0** [comment]
Someone has merged his pull request by himself I destroyed my last commit work what is going on here 😥

**Comment by mmubarak0** [comment]
This going to be chaos now I am out🎆🧨

**Pull request closed without merging**